File: python/calculator.py
import tkinter as tk
from tkinter import ttk
import math

class Calculator:
    def __init__(self, root):
        self.root = root
        self.root.title("Multi-Function Calculator")
        self.root.geometry("400x500")
        self.root.resizable(False, False)
        
        # Create tabs for different calculator functions
        self.tab_control = ttk.Notebook(root)
        
        # Standard calculator tab
        self.standard_tab = ttk.Frame(self.tab_control)
        self.tab_control.add(self.standard_tab, text="Standard")
        
        # Conversion calculator tab
        self.conversion_tab = ttk.Frame(self.tab_control)
        self.tab_control.add(self.conversion_tab, text="Length Conversion")
        
        # Log calculator tab
        self.log_tab = ttk.Frame(self.tab_control)
        self.tab_control.add(self.log_tab, text="Logarithm")
        
        self.tab_control.pack(expand=1, fill="both")
        
        # Setup each calculator interface
        self.setup_standard_calculator()
        self.setup_conversion_calculator()
        self.setup_log_calculator()
        
        # Bind keyboard events for the calculator
        self.root.bind("<Key>", self.handle_keyboard_input)
        
        # Set focus to the main window to capture keyboard inputs
        self.root.focus_set()
    
    def setup_standard_calculator(self):
        # Variables for standard calculator
        self.current_expression = ""
        self.total_expression = ""
        
        # Create display frames
        self.display_frame = self.create_display_frame(self.standard_tab)
        
        # Create display labels
        self.total_expression_label = self.create_display_label(self.display_frame, self.total_expression, position="top")
        self.current_expression_label = self.create_display_label(self.display_frame, self.current_expression, position="bottom")
        
        # Create buttons frame
        self.buttons_frame = self.create_buttons_frame(self.standard_tab)
        
        # Create digit buttons
        self.digits = {
            7: (1, 1), 8: (1, 2), 9: (1, 3),
            4: (2, 1), 5: (2, 2), 6: (2, 3),
            1: (3, 1), 2: (3, 2), 3: (3, 3),
            0: (4, 2), '.': (4, 1)
        }
        self.create_digit_buttons()
        
        # Create operation buttons
        self.operations = {"/": "÷", "*": "×", "-": "-", "+": "+"}
        self.create_operator_buttons()
        
        # Create special buttons
        self.create_special_buttons()
        
        # Configure the grid layout
        self.buttons_frame.rowconfigure(0, weight=1)
        for x in range(1, 5):
            self.buttons_frame.rowconfigure(x, weight=1)
            self.buttons_frame.columnconfigure(x, weight=1)
    
    def setup_conversion_calculator(self):
        # Create frame for conversion calculator
        conversion_frame = ttk.Frame(self.conversion_tab, padding="10")
        conversion_frame.pack(fill="both", expand=True)
        
        # Create input field for value
        ttk.Label(conversion_frame, text="Enter Value:").grid(column=0, row=0, sticky=tk.W, pady=5)
        self.conversion_value = tk.StringVar()
        ttk.Entry(conversion_frame, width=15, textvariable=self.conversion_value).grid(column=1, row=0, pady=5)
        
        # Create conversion type selection
        ttk.Label(conversion_frame, text="Convert From:").grid(column=0, row=1, sticky=tk.W, pady=5)
        self.from_unit = tk.StringVar()
        
        # Create conversion type selection
        ttk.Label(conversion_frame, text="Convert To:").grid(column=0, row=2, sticky=tk.W, pady=5)
        self.to_unit = tk.StringVar()
        
        # Create dropdown lists for unit selection
        length_units = ["Meters", "Centimeters", "Kilometers", "Inches", "Feet", "Miles"]
        
        from_dropdown = ttk.Combobox(conversion_frame, textvariable=self.from_unit, values=length_units, width=12)
        from_dropdown.grid(column=1, row=1, pady=5)
        from_dropdown.current(0)
        
        to_dropdown = ttk.Combobox(conversion_frame, textvariable=self.to_unit, values=length_units, width=12)
        to_dropdown.grid(column=1, row=2, pady=5)
        to_dropdown.current(1)
        
        # Create convert button
        convert_button = ttk.Button(conversion_frame, text="Convert", command=self.perform_conversion)
        convert_button.grid(column=0, row=3, columnspan=2, pady=10)
        
        # Create result display
        ttk.Label(conversion_frame, text="Result:").grid(column=0, row=4, sticky=tk.W, pady=5)
        self.conversion_result = tk.StringVar()
        result_display = ttk.Entry(conversion_frame, textvariable=self.conversion_result, state="readonly", width=20)
        result_display.grid(column=1, row=4, pady=5)

    def setup_log_calculator(self):
        # Create frame for log calculator
        log_frame = ttk.Frame(self.log_tab, padding="10")
        log_frame.pack(fill="both", expand=True)
        
        # Create input field for value
        ttk.Label(log_frame, text="Enter Value:").grid(column=0, row=0, sticky=tk.W, pady=5)
        self.log_value = tk.StringVar()
        ttk.Entry(log_frame, width=15, textvariable=self.log_value).grid(column=1, row=0, pady=5)
        
        # Create log type selection
        ttk.Label(log_frame, text="Log Type:").grid(column=0, row=1, sticky=tk.W, pady=5)
        self.log_type = tk.StringVar()
        
        log_types = ["Natural Log (ln)", "Log base 10 (log10)", "Log base 2 (log2)"]
        log_dropdown = ttk.Combobox(log_frame, textvariable=self.log_type, values=log_types, width=15)
        log_dropdown.grid(column=1, row=1, pady=5)
        log_dropdown.current(0)
        
        # Create calculate button
        calc_button = ttk.Button(log_frame, text="Calculate", command=self.perform_log_calculation)
        calc_button.grid(column=0, row=2, columnspan=2, pady=10)
        
        # Create result display
        ttk.Label(log_frame, text="Result:").grid(column=0, row=3, sticky=tk.W, pady=5)
        self.log_result = tk.StringVar()
        result_display = ttk.Entry(log_frame, textvariable=self.log_result, state="readonly", width=20)
        result_display.grid(column=1, row=3, pady=5)
    
    def create_display_frame(self, parent):
        frame = ttk.Frame(parent)
        frame.pack(expand=True, fill="both", padx=10, pady=10)
        return frame
    
    def create_display_label(self, parent, text, position):
        if position == "top":
            label = ttk.Label(parent, text=text, anchor=tk.E, background="#F8F8F8", font=("Arial", 12))
            label.pack(expand=True, fill="both")
        else:
            label = ttk.Label(parent, text=text, anchor=tk.E, background="#FFFFFF", font=("Arial", 24, "bold"))
            label.pack(expand=True, fill="both")
        return label
    
    def create_buttons_frame(self, parent):
        frame = ttk.Frame(parent)
        frame.pack(expand=True, fill="both", padx=10, pady=10)
        return frame
    
    def create_digit_buttons(self):
        for digit, grid_value in self.digits.items():
            button = ttk.Button(self.buttons_frame, text=str(digit),
                                command=lambda x=digit: self.add_to_expression(x))
            button.grid(row=grid_value[0], column=grid_value[1], sticky=tk.NSEW, padx=2, pady=2)
    
    def create_operator_buttons(self):
        i = 0
        for operator, symbol in self.operations.items():
            button = ttk.Button(self.buttons_frame, text=symbol,
                                command=lambda x=operator: self.append_operator(x))
            button.grid(row=i, column=4, sticky=tk.NSEW, padx=2, pady=2)
            i += 1
    
    def create_special_buttons(self):
        self.create_clear_button()
        self.create_equals_button()
        self.create_percent_button()
        self.create_clear_entry_button()
    
    def create_clear_button(self):
        button = ttk.Button(self.buttons_frame, text="C",
                           command=self.clear)
        button.grid(row=0, column=1, sticky=tk.NSEW, padx=2, pady=2)
    
    def create_clear_entry_button(self):
        button = ttk.Button(self.buttons_frame, text="CE",
                           command=self.clear_entry)
        button.grid(row=0, column=2, sticky=tk.NSEW, padx=2, pady=2)
    
    def create_equals_button(self):
        button = ttk.Button(self.buttons_frame, text="=",
                           command=self.evaluate)
        button.grid(row=4, column=3, columnspan=2, sticky=tk.NSEW, padx=2, pady=2)
    
    def create_percent_button(self):
        button = ttk.Button(self.buttons_frame, text="%",
                           command=self.calculate_percent)
        button.grid(row=0, column=3, sticky=tk.NSEW, padx=2, pady=2)
    
    def add_to_expression(self, value):
        self.current_expression += str(value)
        self.update_current_label()
    
    def append_operator(self, operator):
        self.current_expression += operator
        self.total_expression += self.current_expression
        self.current_expression = ""
        self.update_total_label()
        self.update_current_label()
    
    def clear(self):
        self.current_expression = ""
        self.total_expression = ""
        self.update_total_label()
        self.update_current_label()
    
    def clear_entry(self):
        self.current_expression = ""
        self.update_current_label()
    
    def calculate_percent(self):
        if self.current_expression:
            try:
                # If there's a base value to calculate percentage from
                if self.total_expression and self.total_expression[-1] in "+-*/":
                    base_value = eval(self.total_expression[:-1])
                    percentage = float(self.current_expression) / 100
                    self.current_expression = str(base_value * percentage)
                else:
                    # Just convert to percentage
                    self.current_expression = str(float(self.current_expression) / 100)
                self.update_current_label()
            except Exception as e:
                self.current_expression = "Error"
                self.update_current_label()
    
    def evaluate(self):
        self.total_expression += self.current_expression
        self.update_total_label()
        try:
            self.current_expression = str(eval(self.total_expression))
            self.total_expression = ""
        except Exception as e:
            self.current_expression = "Error"
        finally:
            self.update_current_label()
    
    def update_total_label(self):
        expression = self.total_expression
        for operator, symbol in self.operations.items():
            expression = expression.replace(operator, f' {symbol} ')
        self.total_expression_label.config(text=expression)
    
    def update_current_label(self):
        self.current_expression_label.config(text=self.current_expression[:11])
    
    def perform_conversion(self):
        try:
            value = float(self.conversion_value.get())
            from_unit = self.from_unit.get()
            to_unit = self.to_unit.get()
            
            # Convert to meters first (base unit)
            meters = self.convert_to_meters(value, from_unit)
            
            # Convert from meters to target unit
            result = self.convert_from_meters(meters, to_unit)
            
            self.conversion_result.set(f"{result:.6g}")
        except ValueError:
            self.conversion_result.set("Invalid input")
    
    def convert_to_meters(self, value, unit):
        if unit == "Meters":
            return value
        elif unit == "Centimeters":
            return value / 100
        elif unit == "Kilometers":
            return value * 1000
        elif unit == "Inches":
            return value * 0.0254
        elif unit == "Feet":
            return value * 0.3048
        elif unit == "Miles":
            return value * 1609.34
    
    def convert_from_meters(self, meters, unit):
        if unit == "Meters":
            return meters
        elif unit == "Centimeters":
            return meters * 100
        elif unit == "Kilometers":
            return meters / 1000
        elif unit == "Inches":
            return meters / 0.0254
        elif unit == "Feet":
            return meters / 0.3048
        elif unit == "Miles":
            return meters / 1609.34
    
    def perform_log_calculation(self):
        try:
            value = float(self.log_value.get())
            log_type = self.log_type.get()
            
            if value <= 0:
                self.log_result.set("Invalid input (must be > 0)")
                return
            
            if log_type == "Natural Log (ln)":
                result = math.log(value)
            elif log_type == "Log base 10 (log10)":
                result = math.log10(value)
            else:  # Log base 2
                result = math.log2(value)
            
            self.log_result.set(f"{result:.8g}")
        except ValueError:
            self.log_result.set("Invalid input")
    
    # Add keyboard handler method
    def handle_keyboard_input(self, event):
        # Handle numeric keys and decimal point
        if event.char.isdigit() or event.char == '.':
            self.add_to_expression(event.char)
        # Handle operators
        elif event.char in ['+', '-', '*', '/']:
            self.append_operator(event.char)
        # Handle equals sign and Enter key
        elif event.char == '=' or event.keysym == 'Return':
            self.evaluate()
        # Handle Escape key for clear
        elif event.keysym == 'Escape':
            self.clear()
        # Handle Backspace for clear entry
        elif event.keysym == 'BackSpace':
            if self.current_expression:
                self.current_expression = self.current_expression[:-1]
                self.update_current_label()
        # Handle percent key
        elif event.char == '%':
            self.calculate_percent()

# Run the application
if __name__ == "__main__":
    root = tk.Tk()
    calculator = Calculator(root)
    root.mainloop()

File: java/Calculator.java
import java.awt.*;
import java.awt.event.*;
import java.util.HashMap;
import javax.swing.*;

public class Calculator extends JFrame {
    
    private final JTabbedPane tabbedPane;
    
    // Standard calculator components
    private JPanel standardPanel;
    private JTextField displayField;
    private JLabel expressionLabel;
    private String currentExpression = "";
    private String totalExpression = "";
    private final HashMap<String, String> operations;
    
    // Length conversion components
    private JPanel conversionPanel;
    private JTextField conversionValueField;
    private JComboBox<String> fromUnitCombo;
    private JComboBox<String> toUnitCombo;
    private JTextField conversionResultField;
    
    // Logarithm components
    private JPanel logPanel;
    private JTextField logValueField;
    private JComboBox<String> logTypeCombo;
    private JTextField logResultField;
    
    public Calculator() {
        // Set up the JFrame
        super("Multi-Function Calculator");
        setSize(400, 500);
        setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
        setResizable(false);
        setLocationRelativeTo(null);
        
        // Initialize operations map
        operations = new HashMap<>();
        operations.put("/", "÷");
        operations.put("*", "×");
        operations.put("-", "-");
        operations.put("+", "+");
        
        // Create a tabbed pane
        tabbedPane = new JTabbedPane();
        tabbedPane.setFont(new Font("Arial", Font.BOLD, 12));
        tabbedPane.setBackground(new Color(240, 240, 240));
        
        // Set up tabs
        setupStandardCalculator();
        setupConversionCalculator();
        setupLogarithmCalculator();
        
        // Add tabs to the tabbed pane
        tabbedPane.addTab("Standard", standardPanel);
        tabbedPane.addTab("Length Conversion", conversionPanel);
        tabbedPane.addTab("Logarithm", logPanel);
        
        // Add the tabbed pane to the JFrame
        add(tabbedPane);
        
        // Add keyboard listener for the standard calculator
        addKeyListener(new KeyAdapter() {
            @Override
            public void keyPressed(KeyEvent e) {
                handleKeyboardInput(e);
            }
        });
        
        // Make the frame focusable to receive keyboard events
        setFocusable(true);
        requestFocus();
        
        // Add window focus listener to ensure keyboard focus
        addWindowFocusListener(new WindowAdapter() {
            @Override
            public void windowGainedFocus(WindowEvent e) {
                requestFocus();
            }
        });
    }
    
    private void setupStandardCalculator() {
        standardPanel = new JPanel();
        standardPanel.setLayout(new BorderLayout());
        standardPanel.setBackground(new Color(240, 240, 240));
        
        // Create display panel with a nicer look
        JPanel displayPanel = new JPanel(new BorderLayout(0, 5));
        displayPanel.setBackground(new Color(240, 240, 240));
        
        expressionLabel = new JLabel("");
        expressionLabel.setHorizontalAlignment(JLabel.RIGHT);
        expressionLabel.setFont(new Font("Arial", Font.PLAIN, 14));
        expressionLabel.setBorder(BorderFactory.createEmptyBorder(5, 10, 0, 10));
        expressionLabel.setForeground(new Color(100, 100, 100));
        
        displayField = new JTextField();
        displayField.setHorizontalAlignment(JTextField.RIGHT);
        displayField.setFont(new Font("Arial", Font.BOLD, 28));
        displayField.setEditable(false);
        displayField.setBorder(BorderFactory.createCompoundBorder(
            BorderFactory.createLineBorder(new Color(200, 200, 200), 1, true),
            BorderFactory.createEmptyBorder(8, 10, 8, 10)
        ));
        displayField.setBackground(Color.WHITE);
        
        displayPanel.add(expressionLabel, BorderLayout.NORTH);
        displayPanel.add(displayField, BorderLayout.CENTER);
        displayPanel.setBorder(BorderFactory.createEmptyBorder(15, 10, 10, 10));
        
        standardPanel.add(displayPanel, BorderLayout.NORTH);
        
        // Create buttons panel with improved styling
        JPanel buttonsPanel = new JPanel(new GridLayout(5, 4, 8, 8));
        buttonsPanel.setBorder(BorderFactory.createEmptyBorder(10, 10, 15, 10));
        buttonsPanel.setBackground(new Color(240, 240, 240));
        
        // Create styled buttons
        JButton clearButton = createStyledButton("C", new Color(255, 150, 150));
        clearButton.addActionListener(e -> clearAll());
        
        JButton clearEntryButton = createStyledButton("CE", new Color(255, 200, 150));
        clearEntryButton.addActionListener(e -> clearEntry());
        
        JButton percentButton = createStyledButton("%", new Color(220, 220, 255));
        percentButton.addActionListener(e -> calculatePercent());
        
        // Add buttons for row 1
        buttonsPanel.add(clearButton);
        buttonsPanel.add(clearEntryButton);
        buttonsPanel.add(percentButton);
        buttonsPanel.add(createOperatorButton("/"));
        
        // Add digit buttons
        addDigitButton(buttonsPanel, "7");
        addDigitButton(buttonsPanel, "8");
        addDigitButton(buttonsPanel, "9");
        buttonsPanel.add(createOperatorButton("*"));
        
        addDigitButton(buttonsPanel, "4");
        addDigitButton(buttonsPanel, "5");
        addDigitButton(buttonsPanel, "6");
        buttonsPanel.add(createOperatorButton("-"));
        
        addDigitButton(buttonsPanel, "1");
        addDigitButton(buttonsPanel, "2");
        addDigitButton(buttonsPanel, "3");
        buttonsPanel.add(createOperatorButton("+"));
        
        addDigitButton(buttonsPanel, ".");
        addDigitButton(buttonsPanel, "0");
        
        JButton equalsButton = createStyledButton("=", new Color(120, 200, 255));
        equalsButton.addActionListener(e -> evaluate());
        buttonsPanel.add(equalsButton);
        buttonsPanel.add(new JLabel()); // Empty cell
        
        standardPanel.add(buttonsPanel, BorderLayout.CENTER);
    }
    
    // Create a styled button with custom background
    private JButton createStyledButton(String text, Color bgColor) {
        JButton button = new JButton(text);
        button.setFont(new Font("Arial", Font.BOLD, 16));
        button.setBackground(bgColor);
        button.setForeground(Color.BLACK);
        button.setFocusPainted(false);
        button.setBorder(BorderFactory.createCompoundBorder(
            BorderFactory.createLineBorder(Color.GRAY, 1, true),
            BorderFactory.createEmptyBorder(8, 8, 8, 8)
        ));
        return button;
    }
    
    private void addDigitButton(JPanel panel, String digit) {
        JButton button = createStyledButton(digit, new Color(250, 250, 250));
        button.addActionListener(e -> addToExpression(digit));
        panel.add(button);
    }
    
    private JButton createOperatorButton(String operator) {
        JButton button = createStyledButton(operations.get(operator), new Color(230, 230, 255));
        button.addActionListener(e -> appendOperator(operator));
        return button;
    }
    
    // Keyboard input handler
    private void handleKeyboardInput(KeyEvent e) {
        // Only handle keyboard input when the standard calculator tab is selected
        if (tabbedPane.getSelectedIndex() == 0) {
            char keyChar = e.getKeyChar();
            int keyCode = e.getKeyCode();
            
            // Handle digits and decimal
            if (Character.isDigit(keyChar) || keyChar == '.') {
                addToExpression(String.valueOf(keyChar));
            }
            // Handle operators
            else if (keyChar == '+' || keyChar == '-' || keyChar == '*' || keyChar == '/') {
                appendOperator(String.valueOf(keyChar));
            }
            // Handle enter or equals key
            else if (keyChar == '=' || keyCode == KeyEvent.VK_ENTER) {
                evaluate();
            }
            // Handle escape key for clear
            else if (keyCode == KeyEvent.VK_ESCAPE) {
                clearAll();
            }
            // Handle backspace for removing last character
            else if (keyCode == KeyEvent.VK_BACK_SPACE) {
                if (!currentExpression.isEmpty()) {
                    currentExpression = currentExpression.substring(0, currentExpression.length() - 1);
                    displayField.setText(currentExpression);
                }
            }
            // Handle percent key
            else if (keyChar == '%') {
                calculatePercent();
            }
        }
    }
    
    // Add key listeners to the conversion and logarithm input fields
    private void setupConversionCalculator() {
        conversionPanel = new JPanel();
        conversionPanel.setLayout(new BorderLayout());
        
        JPanel contentPanel = new JPanel();
        contentPanel.setLayout(new GridBagLayout());
        GridBagConstraints gbc = new GridBagConstraints();
        gbc.insets = new Insets(5, 5, 5, 5);
        gbc.fill = GridBagConstraints.HORIZONTAL;
        
        // Value input
        gbc.gridx = 0;
        gbc.gridy = 0;
        contentPanel.add(new JLabel("Enter Value:"), gbc);
        
        gbc.gridx = 1;
        conversionValueField = new JTextField(15);
        contentPanel.add(conversionValueField, gbc);
        
        // From unit selection
        gbc.gridx = 0;
        gbc.gridy = 1;
        contentPanel.add(new JLabel("Convert From:"), gbc);
        
        gbc.gridx = 1;
        String[] lengthUnits = {"Meters", "Centimeters", "Kilometers", "Inches", "Feet", "Miles"};
        fromUnitCombo = new JComboBox<>(lengthUnits);
        contentPanel.add(fromUnitCombo, gbc);
        
        // To unit selection
        gbc.gridx = 0;
        gbc.gridy = 2;
        contentPanel.add(new JLabel("Convert To:"), gbc);
        
        gbc.gridx = 1;
        toUnitCombo = new JComboBox<>(lengthUnits);
        toUnitCombo.setSelectedIndex(1); // Default to centimeters
        contentPanel.add(toUnitCombo, gbc);
        
        // Convert button
        gbc.gridx = 0;
        gbc.gridy = 3;
        gbc.gridwidth = 2;
        gbc.fill = GridBagConstraints.NONE;
        gbc.anchor = GridBagConstraints.CENTER;
        JButton convertButton = new JButton("Convert");
        convertButton.addActionListener(e -> performConversion());
        contentPanel.add(convertButton, gbc);
        
        // Result display
        gbc.gridx = 0;
        gbc.gridy = 4;
        gbc.gridwidth = 1;
        gbc.fill = GridBagConstraints.HORIZONTAL;
        contentPanel.add(new JLabel("Result:"), gbc);
        
        gbc.gridx = 1;
        conversionResultField = new JTextField(15);
        conversionResultField.setEditable(false);
        contentPanel.add(conversionResultField, gbc);
        
        conversionPanel.add(contentPanel, BorderLayout.NORTH);
        
        // Add keyboard event for conversion input
        conversionValueField.addKeyListener(new KeyAdapter() {
            @Override
            public void keyPressed(KeyEvent e) {
                if (e.getKeyCode() == KeyEvent.VK_ENTER) {
                    performConversion();
                }
            }
        });
    }
    
    private void setupLogarithmCalculator() {
        logPanel = new JPanel();
        logPanel.setLayout(new BorderLayout());
        
        JPanel contentPanel = new JPanel();
        contentPanel.setLayout(new GridBagLayout());
        GridBagConstraints gbc = new GridBagConstraints();
        gbc.insets = new Insets(5, 5, 5, 5);
        gbc.fill = GridBagConstraints.HORIZONTAL;
        
        // Value input
        gbc.gridx = 0;
        gbc.gridy = 0;
        contentPanel.add(new JLabel("Enter Value:"), gbc);
        
        gbc.gridx = 1;
        logValueField = new JTextField(15);
        contentPanel.add(logValueField, gbc);
        
        // Log type selection
        gbc.gridx = 0;
        gbc.gridy = 1;
        contentPanel.add(new JLabel("Log Type:"), gbc);
        
        gbc.gridx = 1;
        String[] logTypes = {"Natural Log (ln)", "Log base 10 (log10)", "Log base 2 (log2)"};
        logTypeCombo = new JComboBox<>(logTypes);
        contentPanel.add(logTypeCombo, gbc);
        
        // Calculate button
        gbc.gridx = 0;
        gbc.gridy = 2;
        gbc.gridwidth = 2;
        gbc.fill = GridBagConstraints.NONE;
        gbc.anchor = GridBagConstraints.CENTER;
        JButton calculateButton = new JButton("Calculate");
        calculateButton.addActionListener(e -> performLogCalculation());
        contentPanel.add(calculateButton, gbc);
        
        // Result display
        gbc.gridx = 0;
        gbc.gridy = 3;
        gbc.gridwidth = 1;
        gbc.fill = GridBagConstraints.HORIZONTAL;
        contentPanel.add(new JLabel("Result:"), gbc);
        
        gbc.gridx = 1;
        logResultField = new JTextField(15);
        logResultField.setEditable(false);
        contentPanel.add(logResultField, gbc);
        
        logPanel.add(contentPanel, BorderLayout.NORTH);
        
        // Add keyboard event for log input
        logValueField.addKeyListener(new KeyAdapter() {
            @Override
            public void keyPressed(KeyEvent e) {
                if (e.getKeyCode() == KeyEvent.VK_ENTER) {
                    performLogCalculation();
                }
            }
        });
    }
    
    private void addToExpression(String value) {
        currentExpression += value;
        displayField.setText(currentExpression);
    }
    
    private void appendOperator(String operator) {
        if (!currentExpression.isEmpty()) {
            totalExpression += currentExpression + operator;
            currentExpression = "";
            expressionLabel.setText(formatExpression(totalExpression));
            displayField.setText("");
        } else if (!totalExpression.isEmpty()) {
            // Replace the last operator if there's no current expression
            totalExpression = totalExpression.substring(0, totalExpression.length() - 1) + operator;
            expressionLabel.setText(formatExpression(totalExpression));
        }
    }
    
    private String formatExpression(String expression) {
        for (String op : operations.keySet()) {
            expression = expression.replace(op, " " + operations.get(op) + " ");
        }
        return expression;
    }
    
    private void clearAll() {
        currentExpression = "";
        totalExpression = "";
        expressionLabel.setText("");
        displayField.setText("");
    }
    
    private void clearEntry() {
        currentExpression = "";
        displayField.setText("");
    }
    
    private void evaluate() {
        if (!currentExpression.isEmpty()) {
            try {
                totalExpression += currentExpression;
                expressionLabel.setText(formatExpression(totalExpression));
                
                // Use our custom expression evaluator instead of JavaScript engine
                double result = evaluateExpression(totalExpression);
                
                currentExpression = String.valueOf(result);
                totalExpression = "";
                displayField.setText(currentExpression);
            } catch (RuntimeException e) {
                displayField.setText("Error");
                System.out.println("Calculation error: " + e.getMessage());
                currentExpression = "";
                totalExpression = "";
            }
        }
    }
    
    // Add this custom expression evaluator method
    private double evaluateExpression(String expression) {
        return new ExpressionEvaluator().evaluate(expression);
    }
    
    private void calculatePercent() {
        if (!currentExpression.isEmpty()) {
            try {
                if (!totalExpression.isEmpty() && 
                    (totalExpression.endsWith("+") || totalExpression.endsWith("-") || 
                     totalExpression.endsWith("*") || totalExpression.endsWith("/"))) {
                    
                    // Use our custom evaluator instead of JavaScript engine
                    String baseExpr = totalExpression.substring(0, totalExpression.length() - 1);
                    double baseValue = evaluateExpression(baseExpr);
                    
                    double percentage = Double.parseDouble(currentExpression) / 100;
                    double result = baseValue * percentage;
                    
                    currentExpression = String.valueOf(result);
                } else {
                    // Just convert to percentage
                    double value = Double.parseDouble(currentExpression);
                    currentExpression = String.valueOf(value / 100);
                }
                displayField.setText(currentExpression);
            } catch (Exception e) {
                displayField.setText("Error");
                currentExpression = "";
            }
        }
    }
    
    private void performConversion() {
        try {
            double value = Double.parseDouble(conversionValueField.getText());
            String fromUnit = (String) fromUnitCombo.getSelectedItem();
            String toUnit = (String) toUnitCombo.getSelectedItem();
            
            // Convert to meters first
            double meters = convertToMeters(value, fromUnit);
            
            // Convert from meters to target unit
            double result = convertFromMeters(meters, toUnit);
            
            // Format the result with appropriate precision
            conversionResultField.setText(String.format("%.6g", result));
        } catch (NumberFormatException e) {
            conversionResultField.setText("Invalid input");
        }
    }
    
    private double convertToMeters(double value, String unit) {
        switch (unit) {
            case "Meters":
                return value;
            case "Centimeters":
                return value / 100;
            case "Kilometers":
                return value * 1000;
            case "Inches":
                return value * 0.0254;
            case "Feet":
                return value * 0.3048;
            case "Miles":
                return value * 1609.34;
            default:
                return value;
        }
    }
    
    private double convertFromMeters(double meters, String unit) {
        switch (unit) {
            case "Meters":
                return meters;
            case "Centimeters":
                return meters * 100;
            case "Kilometers":
                return meters / 1000;
            case "Inches":
                return meters / 0.0254;
            case "Feet":
                return meters / 0.3048;
            case "Miles":
                return meters / 1609.34;
            default:
                return meters;
        }
    }
    
    private void performLogCalculation() {
        try {
            double value = Double.parseDouble(logValueField.getText());
            String logType = (String) logTypeCombo.getSelectedItem();
            
            if (value <= 0) {
                logResultField.setText("Invalid input (must be > 0)");
                return;
            }
            
            double result;
            switch (logType) {
                case "Natural Log (ln)":
                    result = Math.log(value);
                    break;
                case "Log base 10 (log10)":
                    result = Math.log10(value);
                    break;
                case "Log base 2 (log2)":
                    result = Math.log(value) / Math.log(2);
                    break;
                default:
                    result = Math.log(value);
            }
            
            logResultField.setText(String.format("%.8g", result));
        } catch (NumberFormatException e) {
            logResultField.setText("Invalid input");
        }
    }
    
    public static void main(String[] args) {
        try {
            // Use a better look and feel
            UIManager.setLookAndFeel(UIManager.getSystemLookAndFeelClassName());
            
            // Adjust UI properties for a more modern look
            UIManager.put("Button.arc", 10);
            UIManager.put("Component.arc", 10);
            UIManager.put("ProgressBar.arc", 10);
            UIManager.put("TextComponent.arc", 10);
        } catch (Exception e) {
            e.printStackTrace();
        }
        
        SwingUtilities.invokeLater(() -> {
            Calculator calculator = new Calculator();
            calculator.setVisible(true);
            calculator.requestFocus(); // Request focus for keyboard input
        });
    }
    
    // Add this inner class at the end of the Calculator class
    private class ExpressionEvaluator {
        private int pos = -1;
        private int ch;
        private String expr;
        
        public double evaluate(String expression) {
            this.expr = expression;
            pos = -1;
            nextChar();
            double x = parseExpression();
            if (pos < expr.length()) {
                throw new RuntimeException("Unexpected character: " + (char)ch);
            }
            return x;
        }
        
        private void nextChar() {
            ch = (++pos < expr.length()) ? expr.charAt(pos) : -1;
        }
        
        private boolean eat(int charToEat) {
            while (Character.isWhitespace(ch)) {
                nextChar();
            }
            if (ch == charToEat) {
                nextChar();
                return true;
            }
            return false;
        }
        
        private double parseExpression() {
            double x = parseTerm();
            while (true) {
                if (eat('+')) {
                    x += parseTerm();
                } else if (eat('-')) {
                    x -= parseTerm();
                } else {
                    return x;
                }
            }
        }
        
        private double parseTerm() {
            double x = parseFactor();
            while (true) {
                if (eat('*')) {
                    x *= parseFactor();
                } else if (eat('/')) {
                    x /= parseFactor();
                } else {
                    return x;
                }
            }
        }
        
        private double parseFactor() {
            if (eat('+')) {
                return parseFactor();
            }
            if (eat('-')) {
                return -parseFactor();
            }
            
            double x;
            int startPos = this.pos;
            
            if (eat('(')) {
                x = parseExpression();
                eat(')');
            } else if (Character.isDigit(ch) || ch == '.') {
                // Parse numbers
                StringBuilder sb = new StringBuilder();
                while (Character.isDigit(ch) || ch == '.') {
                    sb.append((char)ch);
                    nextChar();
                }
                x = Double.parseDouble(sb.toString());
            } else {
                throw new RuntimeException("Unexpected character: " + (char)ch);
            }
            
            return x;
        }
    }
}
